breaker/breaker.go

```diff
@@ -0,0 +1,262 @@
+	Successes            uint64
+	// Failures is the total number of failures returned.
+	Failures             uint64
+	// ConsecutiveSuccesses is the number of consecutive successes returned.
+	ConsecutiveSuccesses uint64
+	// ConsecutiveFailures is the number of consecutive failures returned.
+	ConsecutiveFailures  uint64
+}
+
+func (c *Counter) reset() {
+	c.Requests = 0
+	c.Successes = 0
+	c.Failures = 0
+	c.ConsecutiveSuccesses = 0
+	c.ConsecutiveFailures = 0
+}
+
+func (c *Counter) request() {
+	c.Requests++
+}
+
+func (c *Counter) success() {
+	c.Successes++
+	c.ConsecutiveSuccesses++
+	c.ConsecutiveFailures = 0
+}
+
+func (c *Counter) failure() {
+	c.Failures++
+	c.ConsecutiveFailures++
+	c.ConsecutiveSuccesses = 0
+}
+
+// Fuse represents a Breaker fuse used to trip the breaker.
+type Fuse interface {
+	// Trip determines if the Breaker should be tripped.
+	Trip(Counter) bool
+}
+
+// FuseFunc is an adapter allowing to use a function as a Fuse.
+type FuseFunc func(Counter) bool
+
+// Trip determines if the Breaker should be tripped.
+func (f FuseFunc) Trip(c Counter) bool {
+	return f(c)
+}
+
+// ThresholdFuse trips the Breaker when the total number of failures exceeds the given count.
+func ThresholdFuse(count uint64) FuseFunc {
+	return FuseFunc(func(c Counter) bool {
+		return c.Failures > count
+	})
+}
+
+// ConsecutiveFuse trips the Breaker when the consecutive number of failures exceeds the given count.
+func ConsecutiveFuse(count uint64) FuseFunc {
+	return FuseFunc(func(c Counter) bool {
+		return c.ConsecutiveFailures > count
+	})
+}
+
+// RateFuse trips the Breaker when the percentage of failures exceeds the given rate.
+//
+// rate should be between 0 and 100.
+func RateFuse(rate uint64) FuseFunc {
+	return FuseFunc(func(c Counter) bool {
+		return c.Failures/c.Requests*100 > rate
+	})
+}
+
+// OptFunc represents a configuration function for Breaker.
+type OptFunc func(b *Breaker)
+
+// WithSleep sets the time the Breaker stays open for.
+func WithSleep(d time.Duration) OptFunc {
+	return OptFunc(func(b *Breaker) {
+		b.sleep = d
+	})
+}
+
+// WithTestRequests sets the number of test requests allowed when the Breaker is half-open.
+func WithTestRequests(c uint64) OptFunc {
+	return OptFunc(func(b *Breaker) {
+		b.testRequests = c
+	})
+}
+
+// Breaker is a circuit breaker.
+type Breaker struct {
+	fuse         Fuse
+	sleep        time.Duration
+	testRequests uint64
+
+	mu        sync.Mutex
+	state     State
+	counter   Counter
+	openUntil time.Time
+}
+
+// NewBreaker creates a new Breaker.
+func NewBreaker(f Fuse, opts ...OptFunc) *Breaker {
+	b := &Breaker{
+		fuse:         f,
+		sleep:        10 * time.Second,
+		testRequests: 1,
+	}
+
+	for _, opt := range opts {
+		opt(b)
+	}
+
+	return b
+}
+
+// Run runs the given request if the Breaker allows it.
+//
+// Run returns an error immediately is the Breaker rejects the request,
+// otherwise it returns the result of the request.
+func (b *Breaker) Run(req func() error) error {
```

Comment:
There are 3 deferred statements executed in the course of this func. This **might** slow down some high-throughput cases. Maybe it's worth benchmarking it with and without defers?

Reply:
We can tackle this in a second pass